New device: Roidmi

joserebelo opened the pull request:
Adds support for Roidmi FM Transmitters (only tested on the first Roidmi, for now), closes #1174.

joserebelo commented:
It is not quite ready to be merged yet. I'm not sure on the best way to handle some things, especially the initial request for the current led color/fm frequency (https://github.com/Freeyourgadget/Gadgetbridge/commit/b49895009b3fe8879694d59820f4cf9238ca58a3#diff-49d663aee9cf9b2d3ddcfb8c4774d393R41), which can't be done right after the connection is started or no answer is received.

IzzySoft commented:
@joserebelo maybe you could prefix the title by "WIP:" (Work-In-Progress) as long as it's not ready to merge – and remove that prefix when it's ready?

cpfeiffer commented:
Sorry for the conflicts with level_list_device.xml -- the whole file has been removed and is not necessary anymore. We use the icon defined in DeviceType now.

joserebelo commented:
No problem, I have already rebased it on my local branch, and I have added support for the Roidmi 3S.

I have, however, run into a problem. It seems the Roidmi 1 is now recognized as an ID115, not mattering the order in which I create the coordinators. The only solution seems to be commenting out the ID115Coordinator.

The Roidmi doesn't even seem to use BLE, at least nothing shows up on BLExplorer. I haven't debugged this extensively yet though, so it might still be some problem on my side.

joserebelo commented:
Ignore my previous comment, I forgot about the device key being the same, since I was using it before the rebase.

Besides the requests for the current color/frequency/battery voltage, which need to be moved somewhere else, it should be pretty close to done now, as long as you agree with the rest of the implementation. I just need to do a few more tests on the Roidmi 3S.

joserebelo commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/devices/DeviceCoordinator.java:
```diff
@@ -248,4 +248,20 @@
+    /**
+     * Returns the preset colors supported by the device, if any
+     */
+    int[] getColorPresets();
```
Will do.

joserebelo commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/adapter/GBDeviceAdapterv2.java:
```diff
@@ -290,6 +303,100 @@ public void onDismissed(Snackbar snackbar, int event) {
+                            @Override
+                            public void onClick(DialogInterface dialog, int which) {
+                                float frequency = Float.valueOf(input.getText().toString());
+                                frequency = Float.valueOf(String.format(Locale.getDefault(), "%.1f", frequency));
```
Just to limit it to 1 decimal place, since it doesn't make sense to have more. Maybe there was a better way to do it, and maybe the user should be warned?

cpfeiffer replied:
Ah, I see. If you could make that a separate method with a speaking name or add a comment about this, that would be sufficient, IMHO.

joserebelo commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/impl/GBDevice.java:
```diff
@@ -73,11 +74,14 @@ public GBDevice createFromParcel(Parcel source) {
     private String mBusyTask;
     private List<ItemWithDetails> mDeviceInfos;
+    private int mLedColor = 0;
+    private float mFmFrequency = FREQUENCY_UNKNOWN;
```
I thought about it, and It would, yes. But that would mean they would show up on the device info list when the card is expanded. Can I add a boolean flag to make certain infos not?

joserebelo replied:
Yes, that seems appropriate. Also, this is closely related to #1176, there should be some integration between the device actions such as these and the infos.

cpfeiffer replied:
Yes, the device actions should be able to access and use those extra infos.

joserebelo commented on app/src/main/java/nodomain/freeyourgadget/gadgetbridge/service/devices/roidmi/RoidmiSupport.java:
```diff
@@ -0,0 +1,139 @@
+    public boolean connect() {
+        getDeviceIOThread().start();
+
+        // FIXME not like this
```
@cpfeiffer before merging I still need to find a more elegant solution for this, in order to handle an eventual timeout, and also periodically poll the battery voltage.

Do you know if there is any mechanism already built into Gadgetbridge that would allow me to do that?

cpfeiffer replied:
Other devices start sending battery infos automatically when it gets low.

However, we might also get some generic infrastructure for this, where the UI periodically requests a device update for all connected devices. Either when the display is turned on and Gadgetbridge is in the foreground, or really even periodically when the display is on all the time.

The interval for such update requests might be a device specific value specified by the DeviceCoordinator.

What do you mean exactly with a timeout? Automatic reconnection when a timeout occurs?

joserebelo replied:
Here the values are only sent on request. There's some interest in polling the battery voltage, for eventual warnings if it gets too low / too high (possibly with a customizable threshold).

I am not sure if I am comfortable enough with Gadgetbridge's codebase to implement something like this (especially in a battery-efficient / doze-respecting way) but I can give it a shot.

By timeout I mean requesting the FM Frequency / LED Color again if no answer is received. There should be no need to periodically poll them like the battery voltage, since they will not change, but sometimes the device will not answer the requests, especially if they're made right after the initial connection (hence the Handler with the 3~5 second delay, it almost never fails).

cpfeiffer replied:
I've opened #1189 for the device info updating. I think there shouldn't be any issues regarding battery-friendliness, since we would only do this when the display is active or activated (unless you have other requirements).

joserebelo commented:
Pushed some changes and fixes. I still need to confirm if everything is working properly with the Roidmi 3S, since I don't currently have access to it.

cpfeiffer commented:
OK, great! Just say so when you're ready, from my POV this can be merged.
:+1:

joserebelo commented:
Just an update: I had implemented the Roidmi 3 protocol with just the bluetooth logs. I only got access to my friend's Roidmi 3 today, and of course I missed something and it is not working 😅

I will try to fix it eventually when I have the device for a bit more time.

Meanwhile, the Roidmi 1 is fully working and can be merged, with the support for the Roidmi 3 disabled.

@cpfeiffer do you prefer for me to merge the changes from master or can I rebase the PR (avoiding the merge commit)? I ask this because a rebase might require some extra effort to review the rebased commits.

cpfeiffer commented:
Cool, looking forward :-) Rebase is fine for me :+1:

joserebelo commented:
Alright, I think it should be good to go.

cpfeiffer commented:
Excellent, great job! :ok_hand: :tada:

joserebelo commented:
@cpfeiffer damn it, I just noticed a few hours ago that I used a duplicated device id (watch9), but I was without internet access and couldn't warn in time. 

Everything worked when I tested the rebase, probably because the device wasn't in the database yet, so it was recognized correctly.

I will only be able to fix it in a few hours when I get access to a computer again, if someone can or wants to fix it meanwhile.

I am so sorry 😔

cpfeiffer commented:
Don't worry, :shit: happens and thanks for the heads up. Just send a PR when you have time, there's no imminent release in preparation (and I'm away from the computer as well).
